near-sdk-macros/src/core_impl/abi/abi_generator.rs

```diff
@@ -0,0 +1,158 @@
+use crate::{
+    core_impl::{BindgenArgType, SerializerType},
+    ImplItemMethodInfo, MethodType,
+};
+
+use proc_macro2::{Span, TokenStream as TokenStream2};
+use quote::quote;
+use syn::ReturnType;
+
+use super::TypeRegistry;
+
+impl ImplItemMethodInfo {
+    /// Generates ABI struct for this function.
+    ///
+    /// # Example:
+    /// The following function:
+    /// ```ignore
+    /// fn f3(&mut self, arg0: FancyStruct, arg1: u64) -> Result<IsOk, Error> { }
+    /// ```
+    /// will produce this struct:
+    /// ```ignore
+    /// near_sdk::__private::AbiFunction {
+    ///     name: "f3".to_string(),
+    ///     is_view: false,
+    ///     is_init: false,
+    ///     params: vec![
+    ///         near_sdk::__private::AbiParameter {
+    ///             type_id: 0,
+    ///             serialization_type: "json",
+    ///         },
+    ///         near_sdk::__private::AbiParameter {
+    ///             type_id: 1,
+    ///             serialization_type: "json",
+    ///         }
+    ///     ],
+    ///     callbacks: vec![],
+    ///     callbacks_vec: None,
+    ///     result: near_sdk::__private::AbiParameter {
+    ///         type_id: 2,
+    ///         serialization_type: "json",
+    ///     }
+    /// }
+    /// ```
+    /// If args are serialized with Borsh it will not include `#[derive(borsh::BorshSchema)]`.
+    pub fn abi_struct(&self, registry: &mut TypeRegistry) -> TokenStream2 {
+        let function_name_str = self.attr_signature_info.ident.to_string();
+        let is_view = matches!(&self.attr_signature_info.method_type, &MethodType::View);
+        let is_init = matches!(
+            &self.attr_signature_info.method_type,
+            &MethodType::Init | &MethodType::InitIgnoreState
+        );
+        let params: Vec<TokenStream2> = self
+            .attr_signature_info
+            .input_args()
+            .map(|arg| {
+                let type_id = registry.register_type(Box::new(arg.ty.clone()));
+                let serialization_type = abi_serializer_type(&arg.serializer_ty);
+                let arg_name = arg.ident.to_string();
+                quote! {
+                    near_sdk::__private::AbiParameter {
+                        name: #arg_name.to_string(),
+                        type_id: #type_id,
+                        serialization_type: #serialization_type,
+                    }
+                }
+            })
+            .collect();
+        let callbacks: Vec<TokenStream2> = self
+            .attr_signature_info
+            .args
+            .iter()
+            .filter(|arg| matches!(arg.bindgen_ty, BindgenArgType::CallbackArg))
```

Comment:
This handles callback, but does not handle `CallbackResultArg`. Am I missing something with this?

Reply:
Oh, yeah, you are right. I took it from here: https://github.com/near/near-sdk-rs/blob/master/near-sdk-macros/src/core_impl/metadata/metadata_generator.rs#L69, so I guess the previous metadata implementation is not accounting for this case either. Should I fix it in a separate PR or do you think it is inconsequential at this point?

Reply:
I guess it depends what we do with metadata, but my assumption is that it would be deprecated.

As for callback result, will this not have to be handled more specifically than [`ccfba2e` (#831)](https://github.com/near/near-sdk-rs/pull/831/commits/ccfba2e07128d73748f9e04fd881bdfc3927ea8c) to ignore the wrapping `Result` type in the schema?